core/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -792,6 +793,17 @@ private void parseSource(DefaultSearchContext context, SearchSourceBuilder sourc
             }
             context.storedFieldsContext(source.storedFields());
         }
+
+        if (source.collapse() != null) {
```

Comment:
What if we add this logic to `SearchBuilder#collapse(...)` setter? Downside is that Java client users could unset doc values fields, but we can add a validation check for that in `SearchBuilder#rewrite(...)`. I think that this approach would be cleaner than this changes in this file.